File: sdks/besu/interoperation-node-sdk/src/AssetManager.ts
// * SPDX-License-Identifier: Apache-2.0
// */

/**
* This file provides helper functions for interoperability operations.
**/
/** End file docs */
import * as assetLocksPb from "@hyperledger-labs/weaver-protos-js/common/asset_locks_pb";
import * as Web3 from 'web3-utils';
import { Hash } from "./HashFunctions";

// Create an asset exchange agreement structure
function createAssetExchangeAgreementSerialized(assetType: string, assetID: string, locker: string, recipient: string) {
    const assetExchangeAgreement = new assetLocksPb.AssetExchangeAgreement();
    assetExchangeAgreement.setAssettype(assetType);
    assetExchangeAgreement.setId(assetID);
    assetExchangeAgreement.setRecipient(recipient);
    assetExchangeAgreement.setLocker(locker);
    return Buffer.from(assetExchangeAgreement.serializeBinary())
}

function createFungibleAssetExchangeAgreementSerialized(assetType: string, numUnits: number, locker: string, recipient: string) {
    const assetExchangeAgreement = new assetLocksPb.FungibleAssetExchangeAgreement();
    assetExchangeAgreement.setAssettype(assetType);
    assetExchangeAgreement.setNumunits(numUnits);
    assetExchangeAgreement.setRecipient(recipient);
    assetExchangeAgreement.setLocker(locker);
    return Buffer.from(assetExchangeAgreement.serializeBinary())
}


function createHybridAssetExchangeAgreementSerialized(assetType: string, assetID: string, assetData: string, numUnits: number, recipient: string) {
    const assetExchangeAgreement = new assetLocksPb.HybridAssetExchangeAgreement();
    assetExchangeAgreement.setAssettype(assetType);
    assetExchangeAgreement.setId(assetID);
    assetExchangeAgreement.setRecipient(recipient);
    assetExchangeAgreement.setAssetdata(Buffer.from(assetData, "hex"));
    assetExchangeAgreement.setNumunits(numUnits);
    return Buffer.from(assetExchangeAgreement.serializeBinary())
}

// Create an asset lock structure
function createAssetLockInfoSerialized(hash: Hash, expiryTimeSecs: number) {
    const lockInfoHTLC = new assetLocksPb.AssetLockHTLC();
    lockInfoHTLC.setHashmechanism(hash.HASH_MECHANISM);
    lockInfoHTLC.setHashbase64(Buffer.from(hash.getSerializedHashBase64(), 'base64'));
    lockInfoHTLC.setExpirytimesecs(expiryTimeSecs);
    lockInfoHTLC.setTimespec(assetLocksPb.TimeSpec.EPOCH)
    return Buffer.from(lockInfoHTLC.serializeBinary())
}

// Create an asset claim structure
function createAssetClaimInfoSerialized(hash: Hash) {
    const claimInfoHTLC = new assetLocksPb.AssetClaimHTLC();
    claimInfoHTLC.setHashmechanism(hash.HASH_MECHANISM);
    claimInfoHTLC.setHashpreimagebase64(Buffer.from(hash.getPreimage()));
    const claimInfoHTLCSerialized = claimInfoHTLC.serializeBinary();
    const claimInfo = new assetLocksPb.AssetClaim();
    claimInfo.setLockmechanism(assetLocksPb.LockMechanism.HTLC);
    claimInfo.setClaiminfo(claimInfoHTLCSerialized);
    return Buffer.from(claimInfo.serializeBinary())
}

const createHTLC = async (
    assetManagerContract: any,
    tokenContract: any,
    assetType: string,
    assetID: string,
    senderAddress: string,
    recipientAddress: string,
    expiryTimeSecs: number,
    hash: Hash
): Promise<{ result: any }> => {

    if (!assetManagerContract || !tokenContract) {
        console.log("Contract handle not supplied");
        return { result: false };
    }
    if (!recipientAddress) {
        console.log(`Recipient address not supplied ${recipientAddress}`);
        return { result: false };
    }
    const currTimeSecs = Math.floor(Date.now() / 1000);   // Convert epoch milliseconds to seconds
    if (expiryTimeSecs <= currTimeSecs) {
        console.log("Supplied expiry time invalid or in the past: %s; current time: %s", new Date(expiryTimeSecs).toISOString(), new Date(currTimeSecs).toISOString());
        return { result: false };
    }
    var agreement = createAssetExchangeAgreementSerialized(assetType, assetID, "", recipientAddress.slice(2));

    if (!agreement) {
        console.log("Error creating protobuf asset exchange agreement");
        return { result: false };
    }
    const lockInfo = createAssetLockInfoSerialized(hash, expiryTimeSecs);
    // Normal invoke function
    let lockStatus = await assetManagerContract.lockNonFungibleAsset(
        tokenContract.address,
        agreement,
        lockInfo,
        {
            from: senderAddress,
        }
    ).catch(function (e: any) {
        console.log(e);
        console.log("lockNonFungibleAsset threw an error");
        lockStatus = { result: false }
    })
    return { result: lockStatus };
};

const createFungibleHTLC = async (
    assetManagerContract: any,
    tokenContract: any,
    assetType: string,
    assetAmount: number,
    senderAddress: string,
    recipientAddress: string,
    expiryTimeSecs: number,
    hash: Hash
): Promise<{ result: any }> => {

    if (!assetManagerContract || !tokenContract) {
        console.log("Contract handle not supplied");
        return { result: false };
    }
    if (!recipientAddress) {
        console.log(`Recipient address not supplied ${recipientAddress}`);
        return { result: false };
    }
    const currTimeSecs = Math.floor(Date.now() / 1000);   // Convert epoch milliseconds to seconds
    if (expiryTimeSecs <= currTimeSecs) {
        console.log("Supplied expiry time invalid or in the past: %s; current time: %s", new Date(expiryTimeSecs).toISOString(), new Date(currTimeSecs).toISOString());
        return { result: false };
    }
    var agreement = createFungibleAssetExchangeAgreementSerialized(assetType, assetAmount, "", recipientAddress.slice(2));

    if (!agreement) {
        console.log("Error creating protobuf fungible asset exchange agreement");
        return { result: false };
    }
    const lockInfo = createAssetLockInfoSerialized(hash, expiryTimeSecs);
    // Normal invoke function
    let lockStatus = await assetManagerContract.lockFungibleAsset(
        tokenContract.address,
        agreement,
        lockInfo,
        {
            from: senderAddress,
        }
    ).catch(function (e: any) {
        console.log(e);
        console.log("lockFungibleAsset threw an error");
        lockStatus = { result: false }
    })
    return { result: lockStatus };
};

const createHybridHTLC = async (
    assetManagerContract: any,
    tokenContract: any,
    assetType: string,
    assetID: string,
    assetData: string,
    assetAmount: number,
    senderAddress: string,
    recipientAddress: string,
    expiryTimeSecs: number,
    hash: Hash
): Promise<{ result: any }> => {

    if (!assetManagerContract || !tokenContract) {
        console.log("Contract handle not supplied");
        return { result: false };
    }
    if (!recipientAddress) {
        console.log(`Recipient address not supplied ${recipientAddress}`);
        return { result: false };
    }
    const currTimeSecs = Math.floor(Date.now() / 1000);   // Convert epoch milliseconds to seconds
    if (expiryTimeSecs <= currTimeSecs) {
        console.log("Supplied expiry time invalid or in the past: %s; current time: %s", new Date(expiryTimeSecs).toISOString(), new Date(currTimeSecs).toISOString());
        return { result: false };
    }
    var agreement = createHybridAssetExchangeAgreementSerialized(assetType, String(assetID), Web3.utf8ToHex(String(assetData)), assetAmount, recipientAddress.slice(2));
    if (!agreement) {
        console.log("Error creating protobuf hybrid asset exchange agreement");
        return { result: false };
    }
    const lockInfo = createAssetLockInfoSerialized(hash, expiryTimeSecs);
    // Normal invoke function
    let lockStatus = await assetManagerContract.lockHybridAsset(
        tokenContract.address,
        agreement,
        lockInfo,
        {
            from: senderAddress,
        }
    ).catch(function (e: any) {
        console.log(e);
        console.log("lockHybridAsset threw an error");
        lockStatus = { result: false }
    })
    
    return { result: lockStatus };
};

/**
 * Latter step of a Hashed Time Lock Contract
 * - Claim a unique asset instance using a hash preimage
 **/
const claimAssetInHTLC = async (
    assetManagerContract: any,
    lockContractId: string,
    senderAddress: string,
    hash: Hash,
): Promise<any> => {

    if (!assetManagerContract) {
        console.log("Contract Address not supplied");
        return false;
    }

    // const hash = new SHA256()
    // hash.setPreimage(preimage)
    const claimInfoStr = createAssetClaimInfoSerialized(hash);
    // Normal invoke function
    var claimStatus = await assetManagerContract.claimAsset(lockContractId, claimInfoStr, { from: senderAddress }).catch(function (e: any) {
        console.log(e)
        console.log("claimAsset threw an error");
        claimStatus = false
    })

    return claimStatus;
};

const isAssetLockedInHTLC = async (
    assetManagerContract: any,
    lockContractId: string,
): Promise<any> => {

    if (!assetManagerContract) {
        console.log("Contract not supplied");
        return false;
    }

    // Normal invoke function
    var lockStatus = await assetManagerContract.isAssetLocked(lockContractId).catch(function (e: any) {
        console.log(e)
        console.log("isAssetLock threw an error");
        lockStatus = false
    })

    return lockStatus;
};


const reclaimAssetInHTLC = async (
    interopContract: any,
    lockContractId: string,
    sender: string,
): Promise<void> => {
    var unlockStatus = await interopContract.unlockAsset(lockContractId, {
        from: sender
    }).catch(function (e: any) {
        console.log(e)
        console.log("unlockAsset threw an error");
        unlockStatus = false
    })

    return unlockStatus
}

export {
    createAssetExchangeAgreementSerialized,
    createAssetClaimInfoSerialized,
    createHTLC,
    createFungibleHTLC,
    createHybridHTLC,
    claimAssetInHTLC,
    reclaimAssetInHTLC,
    isAssetLockedInHTLC
};

File: weaver/sdks/corda/src/main/kotlin/com/weaver/corda/sdk/AssetManager.kt
/*
 * Copyright IBM Corp. All Rights Reserved.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package com.weaver.corda.sdk;

import arrow.core.Either
import arrow.core.Left
import arrow.core.Right
import org.slf4j.LoggerFactory

import net.corda.core.messaging.CordaRPCOps
import net.corda.core.contracts.CommandData
import net.corda.core.contracts.StateAndRef
import net.corda.core.messaging.startFlow
import net.corda.core.transactions.SignedTransaction
import net.corda.core.identity.Party
import java.lang.Exception
import java.util.*
import com.google.protobuf.ByteString
import com.weaver.corda.app.interop.flows.*
import com.weaver.corda.app.interop.states.AssetExchangeHTLCState

import com.weaver.protos.common.asset_locks.AssetLocks

class AssetManager {
    companion object {
        private val logger = LoggerFactory.getLogger(AssetManager::class.java)

        @JvmStatic
        @JvmOverloads fun createHTLC(
            proxy: CordaRPCOps,
            assetType: String,
            assetId: String,
            recipientParty: String,
            hash: HashFunctions.Hash,
            expiryTimeSecs: Long,
            timeSpec: Int,
            getAssetStateAndRefFlow: String,
            deleteAssetStateCommand: CommandData,
            issuer: Party,
            observers: List<Party> = listOf<Party>()
        ): Either<Error, String> {
            return try {
                AssetManager.logger.debug("Sending asset-lock request to Corda as part of asset-exchange.\n")

                val result = runCatching {
                    val assetAgreement = createAssetExchangeAgreement(assetType, assetId, recipientParty, "")
                    val lockInfo = createAssetLockInfo(hash, timeSpec, expiryTimeSecs)

                    proxy.startFlow(::LockAsset, lockInfo, assetAgreement, getAssetStateAndRefFlow, deleteAssetStateCommand, issuer, observers)
                        .returnValue.get()
                }.fold({
                    it.map { linearId ->
                        AssetManager.logger.debug("Locking asset was successful and the state was stored with linearId $linearId.\n")
                        linearId.toString()
                    }
                }, {
                    Left(Error("Corda Network Error: Error running LockAsset flow: ${it.message}\n"))
                })
                result
            } catch (e: Exception) {
                AssetManager.logger.error("Error locking asset in Corda network: ${e.message}\n")
                Left(Error("Error locking asset in Corda network: ${e.message}"))
            }
        }

        @JvmStatic
        @JvmOverloads fun createFungibleHTLC(
            proxy: CordaRPCOps,
            tokenType: String,
            numUnits: Long,
            recipientParty: String,
            hash: HashFunctions.Hash,
            expiryTimeSecs: Long,
            timeSpec: Int,
            getAssetStateAndRefFlow: String,
            deleteAssetStateCommand: CommandData,
            issuer: Party,
            observers: List<Party> = listOf<Party>()
        ): Either<Error, String> {
            return try {
                AssetManager.logger.debug("Sending fungible asset-lock request to Corda as part of asset-exchange.\n")
                val result = runCatching {
                    
                    val assetAgreement = createFungibleAssetExchangeAgreement(tokenType, numUnits, recipientParty, "")
                    val lockInfo = createAssetLockInfo(hash, timeSpec, expiryTimeSecs)

                    proxy.startFlow(::LockFungibleAsset, lockInfo, assetAgreement, getAssetStateAndRefFlow, deleteAssetStateCommand, issuer, observers)
                        .returnValue.get()
                }.fold({
                    it.map { linearId ->
                        AssetManager.logger.debug("Locking fungible asset was successful and the state was stored with linearId $linearId.\n")
                        linearId.toString()
                    }
                }, {
                    Left(Error("Corda Network Error: Error running LockFungibleAsset flow: ${it.message}\n"))
                })
                result
            } catch (e: Exception) {
                AssetManager.logger.error("Error locking fungible asset in Corda network: ${e.message}\n")
                Left(Error("Error locking fungible asset in Corda network: ${e.message}"))
            }
        }

        @JvmStatic
        @JvmOverloads fun claimAssetInHTLC(
            proxy: CordaRPCOps,
            contractId: String,
            hash: HashFunctions.Hash,
            createAssetStateCommand: CommandData,
            updateAssetStateOwnerFlow: String,
            issuer: Party,
            observers: List<Party> = listOf<Party>()
        ): Either<Error, SignedTransaction> {
            return try {
                AssetManager.logger.debug("Sending asset-claim request to Corda as part of asset-exchange.\n")
                if (hash.getPreimage() == null) {
                    Left(Error("Error: Preimage can not be null for claim"))
                }
                val signedTx = runCatching {
                    
                    val claimInfo: AssetLocks.AssetClaim = createAssetClaimInfo(hash)

                    proxy.startFlow(::ClaimAsset, contractId, claimInfo, createAssetStateCommand, updateAssetStateOwnerFlow, issuer, observers)
                        .returnValue.get()
                }.fold({
                    it.map { retSignedTx ->
                        AssetManager.logger.debug("Claim asset was successful.\n")
                        retSignedTx
                    }
                }, {
                    Left(Error("Corda Network Error: Error running ClaimAsset flow: ${it.message}\n"))
                })
                signedTx
            } catch (e: Exception) {
                AssetManager.logger.error("Error claiming asset in Corda network: ${e.message}\n")
                Left(Error("Error claiming asset in Corda network: ${e.message}"))
            }
        }

        @JvmStatic
        fun reclaimAssetInHTLC(
            proxy: CordaRPCOps,
            contractId: String,
            createAssetStateCommand: CommandData,
            issuer: Party,
            observers: List<Party> = listOf<Party>()
        ): Either<Error, SignedTransaction> {
            return try {
                AssetManager.logger.debug("Sending asset-unlock request to Corda as part of asset-exchange.\n")
                val signedTx = runCatching {
                    
                    proxy.startFlow(::UnlockAsset, contractId, createAssetStateCommand, issuer, observers)
                        .returnValue.get()
                }.fold({
                    it.map { retSignedTx ->
                        AssetManager.logger.debug("Unlock asset was successful.\n")
                        retSignedTx
                    }
                }, {
                    Left(Error("Corda Network Error: Error running UnlockAsset flow: ${it.message}\n"))
                })
                signedTx
            } catch (e: Exception) {
                AssetManager.logger.error("Error unlocking asset in Corda network: ${e.message}\n")
                Left(Error("Error unlocking asset in Corda network: ${e.message}"))
            }
        }

        @JvmStatic
        fun isAssetLockedInHTLC(
            proxy: CordaRPCOps,
            contractId: String
        ): Boolean {
            return try {
                AssetManager.logger.debug("Querying if asset is locked in Corda as part of asset-exchange.\n")
                val isLocked = proxy.startFlow(::IsAssetLockedHTLC, contractId)
                        .returnValue.get()
                isLocked
            } catch (e: Exception) {
                AssetManager.logger.error("Error querying asset lock state in Corda network: ${e.message}\n")
                false
            }
        }

        @JvmStatic
        fun readHTLCStateByContractId(
            proxy: CordaRPCOps,
            contractId: String
        ): Either<Error, StateAndRef<AssetExchangeHTLCState>> {
            return try {
                AssetManager.logger.debug("Querying asset-lock HTLC state from Corda as part of asset-exchange.\n")
                val obj = runCatching {
                    
                    proxy.startFlow(::GetAssetExchangeHTLCStateById, contractId)
                        .returnValue.get()
                }.fold({
                    it.map { retObj ->
                        AssetManager.logger.debug("Querying HTLC state was successful.\n")
                        retObj
                    }
                }, {
                    Left(Error("Corda Network Error: Error in GetAssetExchangeHTLCStateById flow: ${it.message}\n"))
                })
                obj
            } catch (e: Exception) {
                AssetManager.logger.error("Error querying HTLC state from Corda network: ${e.message}\n")
                Left(Error("Error querying HTLC state from Corda network: ${e.message}"))
            }
        }
        
        @JvmStatic
        fun getHTLCHashById(
            proxy: CordaRPCOps,
            contractId: String
        ): String? {
            try {
                AssetManager.logger.debug("Querying asset-lock HTLC Hash from Corda as part of asset-exchange.\n")
                val obj = proxy.startFlow(::GetAssetExchangeHTLCHashById, contractId)
                        .returnValue.get()
                return obj.toString(Charsets.UTF_8)
            } catch (e: Exception) {
                AssetManager.logger.error("Error querying HTLC Hash from Corda network: ${e.message}\n")
                return null
            }
        }
        
        @JvmStatic
        fun getHTLCHashPreImageById(
            proxy: CordaRPCOps,
            contractId: String
        ): String? {
            try {
                AssetManager.logger.debug("Querying asset-lock HTLC Hash PreImage from Corda as part of asset-exchange.\n")
                val obj = proxy.startFlow(::GetAssetExchangeHTLCHashPreImageById, contractId)
                        .returnValue.get()
                return obj.toString(Charsets.UTF_8)
            } catch (e: Exception) {
                AssetManager.logger.error("Error querying HTLC Hash PreImage from Corda network: ${e.message}\n")
                return null
            }
        }

        fun createAssetExchangeAgreement(
            assetType: String,
            assetId: String,
            recipient: String,
            locker: String): AssetLocks.AssetExchangeAgreement {

                val assetAgreement = AssetLocks.AssetExchangeAgreement.newBuilder()
                    .setAssetType(assetType)
                    .setId(assetId)
                    .setLocker(locker)
                    .setRecipient(recipient)
                    .build()

                return assetAgreement
        }

        fun createFungibleAssetExchangeAgreement(
            tokenType: String,
            numUnits: Long,
            recipient: String,
            locker: String): AssetLocks.FungibleAssetExchangeAgreement {
            
                val assetAgreement = AssetLocks.FungibleAssetExchangeAgreement.newBuilder()
                    .setAssetType(tokenType)
                    .setNumUnits(numUnits)
                    .setLocker(locker)
                    .setRecipient(recipient)
                    .build()

                return assetAgreement
        }

        fun createAssetLockInfo(
            hash: HashFunctions.Hash,
            timeSpec: Int,
            expiryTimeSecs: Long): AssetLocks.AssetLock {
	        
            val lockInfoHTLC = AssetLocks.AssetLockHTLC.newBuilder()
                .setHashMechanism(hash.HASH_MECHANISM)
                .setHashBase64(ByteString.copyFrom(hash.getSerializedHashBase64().toByteArray()))
                .setTimeSpecValue(timeSpec)
                .setExpiryTimeSecs(expiryTimeSecs)
                .build()

            val lockInfo = AssetLocks.AssetLock.newBuilder()
                .setLockMechanism(AssetLocks.LockMechanism.HTLC)
                .setLockInfo(ByteString.copyFrom(Base64.getEncoder().encodeToString(lockInfoHTLC.toByteArray()).toByteArray()))
                .build()

            return lockInfo
        }

        fun createAssetClaimInfo(
            hash: HashFunctions.Hash): AssetLocks.AssetClaim {
	        
            val claimInfoHTLC = AssetLocks.AssetClaimHTLC.newBuilder()
                .setHashMechanism(hash.HASH_MECHANISM)
                .setHashPreimageBase64(ByteString.copyFrom(hash.getSerializedPreimageBase64()!!.toByteArray()))
                .build()

            val claimInfo = AssetLocks.AssetClaim.newBuilder()
                .setLockMechanism(AssetLocks.LockMechanism.HTLC)
                .setClaimInfo(ByteString.copyFrom(Base64.getEncoder().encodeToString(claimInfoHTLC.toByteArray()).toByteArray()))
                .build()

            return claimInfo    
        }
    }
}
